service-registry/src/main/java/io/servicecomb/serviceregistry/consumer/MicroserviceVersionRule.java

```diff
@@ -0,0 +1,145 @@
+
+  // key is instanceId
+  private Map<String, MicroserviceInstance> instances = new ConcurrentHashMap<>();
+
+  private InstanceCache instanceCache;
+
+  public MicroserviceVersionRule(String appId, String microserviceName, String strVersionRule) {
+    this.appId = appId;
+    this.microserviceName = microserviceName;
+    this.versionRule = VersionRuleUtils.getOrCreate(strVersionRule);
+
+    resetInstanceCache();
+  }
+
+  private void resetInstanceCache() {
+    instanceCache = new InstanceCache(appId, microserviceName, versionRule.getVersionRule(), instances);
+  }
+
+  public void addMicroserviceVersion(MicroserviceVersion microserviceVersion) {
+    if (!versionRule.isAccept(microserviceVersion.getVersion())) {
+      return;
+    }
+
+    versions.put(microserviceVersion.getMicroservice().getServiceId(), microserviceVersion);
+    resetLatest();
+
+    LOGGER.info("add microserviceVersion, appId={}, microserviceName={}, version={}, versionRule={}.",
+        microserviceVersion.getMicroservice().getAppId(),
+        microserviceVersion.getMicroservice().getServiceName(),
+        microserviceVersion.getVersion().getVersion(),
+        versionRule.getVersionRule());
+  }
+
+  public void deleteMicroserviceVersion(MicroserviceVersion microserviceVersion) {
+    if (!versionRule.isAccept(microserviceVersion.getVersion())) {
+      return;
+    }
+
+    if (versions.remove(microserviceVersion.getMicroservice().getServiceId()) == null) {
+      return;
+    }
+
+    resetLatest();
+    LOGGER.info("delete microserviceVersion, appId={}, microserviceName={}, version={}, versionRule={}.",
+        microserviceVersion.getMicroservice().getAppId(),
+        microserviceVersion.getMicroservice().getServiceName(),
+        microserviceVersion.getVersion().getVersion(),
+        versionRule.getVersionRule());
+  }
+
+  protected void resetLatest() {
+    latestVersion = null;
+    if (versions.isEmpty()) {
+      return;
+    }
+
+    latestVersion = versions.values().stream().sorted((v1, v2) -> {
+      return v2.version.compareTo(v1.version);
+    }).findFirst().get();
+  }
+
+  public VersionRule getVersionRule() {
+    return versionRule;
+  }
+
+  @SuppressWarnings("unchecked")
+  public <T extends MicroserviceVersion> T getLatestMicroserviceVersion() {
+    return (T) latestVersion;
+  }
+
+  public Map<String, MicroserviceInstance> getInstances() {
+    return instances;
+  }
+
+  public InstanceCache getInstanceCache() {
+    return instanceCache;
+  }
+
+  public void setInstances(Collection<MicroserviceInstance> newInstances) {
+    if (newInstances == null) {
+      return;
+    }
+
+    instances = newInstances.stream().filter(instance -> {
+      MicroserviceVersion microserviceVersion = versions.get(instance.getServiceId());
+      return microserviceVersion != null
+          && versionRule.isMatch(microserviceVersion.getVersion(), latestVersion.getVersion());
+    }).filter(instance -> {
```

Comment:
Please use forEach instead of filter.

Reply:
forEach is a "terminal operation"
forEach is not the purpose, Collectors.toMap is the purpose.